### ISSUE TITLE
cannot search word without hashtag

### ISSUE THREAD

**Issue by baptx** [friendica/friendica#9231]
### Expected behavior
It should be possible to search a word that was mentioned by other people without using a hashtag, like on Twitter.

### Actual behavior
It is not possible to search a word if the message is not using a hashtag for the word.

### Steps to reproduce the problem
Create a message with the word Friendica without hashtag on a public server like https://social.isurf.ca/ (direct link to message: https://social.isurf.ca/display/c443a55c-105f-64cb-a7ec-96a959460101).
Try to search the word Friendica (direct link to search results: https://social.isurf.ca/search?q=friendica).

### Friendica version you encountered the problem
2020.09-rc

### Friendica source (git, zip)
https://social.isurf.ca/

### PHP version
N/A

### SQL version
N/A

**Comment by MrPetovan** [comment]
Full-text search can be disabled by node admins because of its terrible performance once the number of stored items is past a threshold that is very easy to reach even on a single-user node. However we should be more clear about this.

**Comment by annando** [comment]
With the newer item table structure we could check whether we could use the fulltext search index from MySQL again.

**Comment by annando** [comment]
Small update: with a fulltext index the search is working really fine, including `and` and `or` stuff. I think we can address this in the next release.

**Comment by baptx** [comment]
@MrPetovan Thanks for the information. Is there a public server using full-text search that you can share? It could be useful to add a tag on https://dir.friendica.social/servers when it is supported, in addition to filters so we can see the list of servers supporting HTTPS or full-text search, without having to check each page.

**Comment by MrPetovan** [comment]
The full text search disabling setting was added because we couldn't figure out a scaling full text index, but if @annando is correct, then we will likely remove this setting soon which we won't need to advertise in the directory anymore.

**Comment by annando** [comment]
Yeah, stuff like this is working fine:

```
MariaDB > select `uid`,`created` from item where `uri-id` in (select `uri-id` from `item-content` where match (body) against ('+björk +album' IN BOOLEAN MODE)) and uid = 0 order by created desc limit 10;
+-----+---------------------+
| uid | created             |
+-----+---------------------+
|   0 | 2020-07-28 06:19:00 |
|   0 | 2020-05-27 11:58:37 |
|   0 | 2017-11-28 21:54:05 |
|   0 | 2016-06-25 22:21:57 |
|   0 | 2016-06-20 20:40:00 |
|   0 | 2016-05-29 09:15:00 |
|   0 | 2014-12-27 20:29:34 |
+-----+---------------------+
7 rows in set (0.02 sec)

MariaDB >
```
And my DB is rather big:

```
MariaDB > select count(*) from `item-content`;
+----------+
| count(*) |
+----------+
|  1317669 |
+----------+
1 row in set (0.67 sec)

MariaDB > 
```

**Comment by AlfredSK** [comment]
I can do a test search. I have a huge DB. FTS takes ages on my main node. That's why it's disabled.
Any suggestions what to search for?

**Comment by AlfredSK** [comment]
Search for 'Kaffee':
Without FT index 94 sec. 😱
With FT index **2** sec.

Search for 'Netflix movie':
Without FT index 55 sec.
With FT index **3** sec.

Search for 'Luke, I am your father.':
Without FT index 36 sec.
With FT index **19** sec. Hmm. That's still a lot.

**Comment by annando** [comment]
@AlfredSK when you perform the tests, please keep in mind that we use the boolean mode of the fulltext search, see https://dev.mysql.com/doc/refman/8.0/en/fulltext-boolean.html

Means: When you want to search for anything where "netflix" and "movie" are mentioned, you type in `+netflix +movie`. When you directly search for a phrase, you enter this `"Luke, I am your father"`. (double quotes).

**Comment by tobiasd** [comment]
I've search for "Bonn" on the RasPi2 ... 42 seconds before the DB update, 4 seconds afterwards.

Is searching for multiple words possible now as well?

**Comment by tobiasd** [comment]
The "+" operator does not work for me. It yields the same results as no operator ("or").

The "-" operator works.

**Comment by AlfredSK** [comment]
I'm getting a lot of Twitter and RSS results from sources I don't follow. Is this right? I always thought that content is only accessible for the user who is following these sources.

**Comment by AlfredSK** [comment]
> when you perform the tests, please keep in mind that we use the boolean mode of the fulltext search

@annando I'm not trying to do searches that are 'optimized' for the way the FTS is implemented. I'm trying to do what a normal human being a.k.a. user would probably put into the search field. ;-)

**Comment by annando** [comment]
Entering multiple words act like `or`.

**Comment by tobiasd** [comment]
Yes, the "or" by default is a bit surprising as well. But the "+" operator is stripped from the search it seems, so there is no possibility to search with "and".

**Comment by annando** [comment]
Can you check if there is a difference between "vier" and "frio"? I tested it and it worked fine here.

**Comment by tobiasd** [comment]
Both frio and vier do strip the "+".

**Comment by AlfredSK** [comment]
Hmm. For me the results are different. `+netflix +umpalumpa` gives no results. `netflix umpalumpa` gives a lot of results. (frio)

**Comment by MrPetovan** [comment]
> Yes, the "or" by default is a bit surprising as well.

The results are ordered by decreasing relevance, meaning that after performing a search on two words, results with both words will show before the results with only either one.

**Comment by tobiasd** [comment]
> The results are ordered by decreasing relevance, meaning that after performing a search on two words, results with both words will show before the results with only either one.

No, that is not the case here (MariaDB 10.3.23). When I search for "Hackathon Berlin" I first get the postings from this evening and afternoon containing "Berlin" and only later my posting which contains both terms.

**Comment by MrPetovan** [comment]
Let me rephrase: the database management software assigns a relevance to results when performing a full-text search. This relevance can be used to sort the results including both terms first. However, on the Friendica side we sort the search results by uri-id, which roughly means by the date it was created in the system.

**Comment by tobiasd** [comment]
That may be the case yes ;-)

Question is why the "+" is removed (search for "+Hackathon +Berlin" and the result page shows "Hackathon Berlin") from the search string on my setup, but not e.g. on libranet. I can copy the URL part from the one tab in the browser to the other and the result stays the same. And it is independent of the used browser.

What is the constructed search query fired onto the database? I'll evaluate if the problem is my DB or the servers rewrite rules.

**Comment by annando** [comment]
The search query should be put directly into the database query. Set your logfile to `info` then you should see it in your log:
```
$ tail -f ../logs/friendica.log | grep -a Plumperquatsch
2020-09-20T07:33:10Z index [INFO]: Start fulltext search. {"q":"+Plumperquatsch"} - {"file":"Index.php","line":157,"function":"content","uid":"f74e44","process_id":15013}
2020-09-20T07:33:11Z index [INFO]: Start Conversation. {"q":"+Plumperquatsch"} - {"file":"Index.php","line":182,"function":"content","uid":"f74e44","process_id":15013}
```

**Comment by tobiasd** [comment]
Then the "+" is stripped before the log entry is written.

> index [INFO]: Start fulltext search. {"q":"hackathon  Berlin"} - {"file":"Index.php","line":157,"function":"content","uid":"286b2d","process_id":29696}

**Comment by annando** [comment]
Oooookay ... weird.

**Comment by tobiasd** [comment]
I'll see what the lighttpd logs are saying once I enable them again in the afternoon. But this problem seems to be a local one, so it should not stop us from the release. :)

**Comment by tobiasd** [comment]
The access log correctly has the line

    GET /search?q=%2BBerlin+%2BHackathon&search-option=fulltext

the `%2B` is the encoded `+` sign. The _routing_ should include the parameters <del>and then the search module</del> (search already lost the `+`) right? ... echoing the `$_GET` variable at the top of the `index.php` already has lost the `+` so it is a lighttpd config issue.

**Comment by MrPetovan** [comment]
Oooh, nasty.

**Comment by tobiasd** [comment]
Yes, I'll keep digging hopefully I'll find a solution :)

**Comment by tobiasd** [comment]
... just FYI the problem with lighttpd and the vanishing operators has been vanished without me finding the time to further investigate it...
